inc/saiprefixcompression.h

```diff
@@ -0,0 +1,305 @@
+/**
+ * Copyright (c) 2024 Microsoft Open Technologies, Inc.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License"); you may
+ *    not use this file except in compliance with the License. You may obtain
+ *    a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR
+ *    CONDITIONS OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT
+ *    LIMITATION ANY IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS
+ *    FOR A PARTICULAR PURPOSE, MERCHANTABILITY OR NON-INFRINGEMENT.
+ *
+ *    See the Apache Version 2.0 License for specific language governing
+ *    permissions and limitations under the License.
+ *
+ *    Microsoft would like to thank the following companies for their review and
+ *    assistance with these files: Intel Corporation, Mellanox Technologies Ltd,
+ *    Dell Products, L.P., Facebook, Inc., Marvell International Ltd.
+ *
+ * @file    saiprefixcompression.h
+ *
+ * @brief   This module defines SAI prefix compression interface
+ */
+
+#if !defined (__SAIPREFIXCOMPRESSION_H_)
+#define __SAIPREFIXCOMPRESSION_H_
+
+#include <saitypes.h>
+
+/**
+ * @defgroup SAIPREFIXCOMPRESSION SAI - Prefix Compression API definitions
+ *
+ * @{
+ */
+
+/**
+ * @brief Attribute Id for SAI prefix compression object
+ */
+typedef enum _sai_prefix_compression_table_attr_t
+{
+    /**
+     * @brief Start of attributes
+     */
+    SAI_PREFIX_COMPRESSION_TABLE_ATTR_START,
+
+    /**
+     * @brief Label attribute used to unique identify Table.
+     *
+     * @type char
+     * @flags CREATE_AND_SET
+     * @default ""
+     */
+    SAI_PREFIX_COMPRESSION_TABLE_ATTR_LABEL = SAI_PREFIX_COMPRESSION_TABLE_ATTR_START,
+
+    /**
+     * @brief Prefix Compression table default entry BINCODE.
+     * BINCODE applied to any address that does not match a prefix in the table
+     *
+     * @type sai_uint32_t
+     * @flags MANDATORY_ON_CREATE | CREATE_AND_SET
+     */
+    SAI_PREFIX_COMPRESSION_TABLE_ATTR_DEFAULT_ENTRY_BINCODE,
+
+    /**
+     * @brief End of attributes
+     */
+    SAI_PREFIX_COMPRESSION_TABLE_ATTR_END,
+
+    /** Custom range base value */
+    SAI_PREFIX_COMPRESSION_TABLE_ATTR_CUSTOM_RANGE_START = 0x10000000,
+
+    /** End of custom range base */
+    SAI_PREFIX_COMPRESSION_TABLE_ATTR_CUSTOM_RANGE_END
+
+} sai_prefix_compression_table_attr_t;
+
+/**
+ * @brief Attribute Id for SAI prefix compression object
+ */
+typedef enum _sai_prefix_compression_entry_attr_t
+{
+    /**
+     * @brief Start of attributes
+     */
+    SAI_PREFIX_COMPRESSION_ENTRY_ATTR_START,
+
+    /**
+     * @brief Prefix Compression entry BINCODE
+     *
+     * @type sai_uint32_t
+     * @flags MANDATORY_ON_CREATE | CREATE_AND_SET
+     */
+    SAI_PREFIX_COMPRESSION_ENTRY_ATTR_BINCODE = SAI_PREFIX_COMPRESSION_ENTRY_ATTR_START,
```

Comment:
It would be better to use META instead of BINCODE to align with other existing SAI attributes.
 SAI_PREFIX_COMPRESSION_ENTRY_ATTR_USER_META

Reply:
I don't mind changing this to META, the reason I used BINCODE because it would be a unique name and won't get confused of the meta in ACL. I don't mind discussing this and if this is a strong ask, then I can change it.

Reply:
BINCODE is NOT aligning to the terms already used in SAI.  Today SAI already uses META for the different qualifiers (route, neighbor, FDB etc) that can be matched in ACL and disambiguates them well. For e.g., SAI_ACL_ENTRY_ATTR_FIELD_SRC_**PREFIX_COMPRESSION**_META clearly disambiguates it from other META that can be matched in the ACL table.

Reply:
I have no problem with that, I can change all BINCODE to META in this PR. Thank you for pointing that out, that is a good point. I will try to have the changes done in the next two days

Reply:
I pushed a change to replace BINCODE with META. Let me know if I can resolve this comment

Reply:
@rck-innovium can you please resolve this if you are ok with the changes I pushed up